plugins/oauth/server/providers/base.py

```diff
@@ -0,0 +1,237 @@
+        """
+        self.clientId = clientId or self.getClientIdSetting()
+        self.clientSecret = clientSecret or self.getClientSecretSetting()
+        self.redirectUri = redirectUri
+
+    @classmethod
+    def getProviderName(cls, external=False):
+        providerName = cls.__name__
+        if external:
+            return providerName
+        else:
+            return providerName.lower()
+
+    def getClientIdSetting(self):
+        raise NotImplementedError()
+
+    def getClientSecretSetting(self):
+        raise NotImplementedError()
+
+    @classmethod
+    def getUrl(cls, state):
+        """
+        Return the URL to start the OAuth2 authentication flow with an external
+        provider.
+
+        This abstract function must be reimplemented by each provider.
+
+        :param state: A unique string, used to prevent CSRF and store other
+        state information. This string should not be URL-encoded (that must be
+        done in this method).
+        :return: An external absolute URL to the start point for this provider's
+        OAuth2 flow.
+        """
+        raise NotImplementedError()
+
+    def getToken(self, code):
+        raise NotImplementedError()
+
+    def getUser(self, token):
+        """
+        Returns a user via this OAuth2 provider, given an OAuth2 access token.
+
+        Subclasses must implement this to retrieve a provider-specific unique
+        identifier, email address, first name, last name, and optionally a
+        login name from the OAuth2 provider, and then call
+        self._createOrReuseUser with that information to lookup or create a
+        corresponding Girder user.
+
+        :param token: The current OAuth2 access token.
+        :returns: The user document corresponding to this user.
+        """
+        raise NotImplementedError()
+
+    def _getJson(self, **kwargs):
+        """
+        Make an HTTP request using the specified kwargs, then parse it as JSON
+        and return the value. If an error occurs, this raises an appropriate
+        exception containing the information.
+        """
+        resp = requests.request(**kwargs)
+        content = resp.content
+
+        if isinstance(content, six.binary_type):
+            content = content.decode('utf8')
+
+        try:
+            resp.raise_for_status()
+        except requests.HTTPError:
+            raise RestException(
+                'Got %s from %s, response="%s".' % (
+                    resp.status_code, kwargs['url'], content
+                ), code=502)
+
+        try:
+            return json.loads(content)
+        except ValueError:
+            raise RestException('Non-JSON response: %s' % content, code=502)
+
+    def _createOrReuseUser(self, oauthId, email, firstName, lastName,
+                           userName=None):
+        providerName = self.getProviderName()
+
+        # Try finding by ID first, since a user can change their email address
+        user = self.model('user').findOne({
```

Comment:
We should add an index to the 'oauth' field. To do this, we should store ``{'provider':..., 'id':....}``, so a compound index can be created.

Thanks @zachmullen .